src/index.ts

```diff
@@ -39,36 +39,26 @@ export type Options = {
  * // When using 1, this will reject if the server isn't already running.
  * untilConnected({ target: 3000, maxRetries: 1 })
  */
-export function untilConnected(options: Options): Promise<void> {
+export async function untilConnected(options: Options): Promise<void> {
   const { port, host } = resolveConnectionOptions(options)
-  let maxRetries = options.maxRetries || 5
-  let retries = 0
+  const maxRetries = options.maxRetries || 5
 
-  function connectOrRetry(): Promise<void> {
-    retries++
-
-    return connect(port, host).catch(async (connectionError) => {
-      if (retries === maxRetries) {
-        const error = new Error(
-          `Failed to await connection at ${
-            host || ''
-          }:${port}. Connection never established (retries: ${retries}).`,
-          { cause: connectionError }
-        )
-        throw error
-      }
-
-      if (options.connectionInterval) {
-        await new Promise<void>((resolve) => {
-          setTimeout(resolve, options.connectionInterval)
-        })
-      }
-
-      return connectOrRetry()
-    })
+  let connectionError = null
+  for (let retry = 0; retry < maxRetries; retry++) {
+    if (retry && options.connectionInterval) {
```

Comment:
`options.connectionInterval != null` may be a better way to check the value here. That way, we'd allow `0` as the value too.

Reply:
We'd have to check against `undefined` instead of `null`. Zero delay means "no delay" whatsoever ;)

Reply:
In any case, just `options.connectionInterval` is incorrect as it ignores `0` as the expected value. `typeof options.connectionInterval !== 'undefined'` is fine by me just bear in mind that will also be `true` for null, which is an incorrect value to `setTimeout`.

Reply:
Since pretty much any value is allowed in `setTimeout`, I've added the `x in options` logic that only checks if the `connectionInterval` was provided at all (even if it's explicit undefined).